### ISSUE TITLE
Upgrade 5.x to 6.x: dark header now flashes a light color during navigation + more

### ISSUE THREAD

**Issue by aforty** [react-navigation/react-navigation#9883]
### Current behavior

I have a fairly complex navigation system in my app with my own custom header. I have simplified the entire thing to a demo (link below) but basically I need to create and apply my own animations.

I use a dark style on my app with a dark background and a dark header bar and there are two issues I've noticed since upgrading to v6.

1. With v6 I can now clearly see the navigation bar flashing a lighter color during navigation when using `HeaderStyleInterpolators.forFade()`. This was not an issue in v5. There is no known work-around for this.
2. Right at the end of the navigation animation for the card itself I can momentarily see a white bar on the left (after the new card slides in from the right). A work-around I've found is setting `detachPreviousScreen: false`. This seems like a bug to me but is there a way for me to not get the flash and still be able to detach previous screens?

https://user-images.githubusercontent.com/523683/131183397-e8b7f953-68d6-49df-9ad1-8ec8926aa79d.mov

### Expected behavior

I would expect me to be able to use whatever color header and background I want without react-navigation momentarily flashing other colors beneath it. Neither of these issues occurred in v5 so I'm strongly leaning towards these being bugs but hopefully I'm mistaken.

### Reproduction

https://snack.expo.dev/@aforty/header-flashes-lighter-during-navigation

### Platform

- [X] Android
- [X] iOS
- [X] Web
- [ ] Windows
- [ ] MacOS

### Packages

- [ ] @react-navigation/bottom-tabs
- [ ] @react-navigation/drawer
- [ ] @react-navigation/material-bottom-tabs
- [ ] @react-navigation/material-top-tabs
- [X] @react-navigation/stack
- [ ] @react-navigation/native-stack

### Environment

| package                                | version |
| -------------------------------------- | ------- |
| @react-navigation/native               | v6.0.6
| @react-navigation/bottom-tabs          | n/a
| @react-navigation/drawer               | n/a
| @react-navigation/material-bottom-tabs | n/a
| @react-navigation/material-top-tabs    | n/a
| @react-navigation/stack                | v6.0.11
| @react-navigation/native-stack         | n/a
| react-native-safe-area-context         | v3.2.0
| react-native-screens                   | v3.6.0
| react-native-gesture-handler           | v1.10.3
| react-native-reanimated                | n/a
| react-native-tab-view                  | n/a
| react-native-pager-view                | n/a
| react-native                           | v0.65.1
| expo                                   | n/a
| node                                   | v16.5.0
| npm or yarn                            | yarn@v7.20.6

**Comment by seanhuggins1** [comment]
Can confirm `detachPreviousScreen: false` works for me as well as a temporary work around

**Comment by yataebi** [comment]
This issue(https://github.com/react-navigation/react-navigation/issues/9089) may be involved.

**Comment by tr3v3r** [comment]
The same here. The issue appeared after upgrading RN to 0.65.1. 
btw we are using react navigation v5

**Comment by cristianoccazinsp** [comment]
Any ideas where the background is coming from? I've tried to mess around with the animation config and styles and it's always white.

**Comment by aforty** [comment]
It's been months now and this is keeping me from upgrading. Any movement on this?

**Comment by pcprinz** [comment]
I had the same issue a while ago and solved the problem by setting the backgroundColor of react navigations theme to my custom backgroundColor

**Comment by aforty** [comment]
@pcprinz You're right, thank you!

Followed the custom section here to add my own background color for light and dark modes: https://reactnavigation.org/docs/themes

**Comment by cristianoccazinsp** [comment]
Still broken, why would it be closed?

**Comment by aforty** [comment]
@cristianoccazinsp See my comment above. Adding a `theme` to the `NavigationContainer` fixes the issue.

**Comment by cristianoccazinsp** [comment]
Sorry, I missed that. I will give it a try to see what happens and report back.

I find it odd that a navigation lib also provides a theming solution though. Most apps will decide on using a separate UI framework which usually comes with its own theme implementation.

**Comment by cristianoccazinsp** [comment]
@aforty months after, I still have the same issue. I have tried using the theme prop, and all it does is to show that color right at the end of the navigation as the poster's bug report. This is of no use, I don't need a color there, the animation should be smooth and not show that kind of artifact.

**Comment by aforty** [comment]
I used the theme prop to provide my specific dark (and light) background color so that's then shown during transitions and it now looks correct.

**Comment by aforty** [comment]
Here's the general gist.

```
import { useColorScheme } from 'react-native';
import { NavigationContainer } from '@react-navigation/native';

const navigationThemes = {
  dark: {
    dark: true,
    colors: {
      background: 'black',
    },
  },
  light: {
    dark: false,
    colors: {
      background: 'white',
    },
  },
};

function Navigator () {
  const scheme = useColorScheme();

  return (
    <NavigationContainer theme={scheme === 'dark' ? navigationThemes.dark : navigationThemes.light}>
      {/* ... */}
    </NavigationContainer>
  );
}
```

**Comment by cristianoccazinsp** [comment]
The problem is that the background color should not show at all, as it does with 5.x.  With 5.x, the screen fully animates to the edge of the screen and only after the screen behind it disappears.

With 6.x, this is not the case, the previous screen is removed maybe 10 pixels before the new screen fully animates into the screen, showing the background for half a second.  It doesn't matter if the background is white or black, either color looks awful as it basically "chops" the previous screen and it's clearly visible if you slow down the transitions.

**Comment by cristianoccazinsp** [comment]
The use case is pretty simple, if I set the background to white, I see a white line at the left of the screen for half a second before the new screen finishes animating. If I set it black, I see the same black line. Neither white nor black lines are visually pleasant and make it look like the app is has visual artifacts.

The main issue is that the app has a persistent header that is re-mounted on each new screen (`headerMode="screen"`}, so this background basically cuts through the entire screen and a very visible white or black line is shown, similar to what you see in your video.

**Comment by dngconsulting** [comment]
Hello,

Have the same problem and all the solutions proposed here are pretty ugly and don't work. Whatever theming colors I set, I get a flash on the header because my header background is ... a linear gradient. The flash appears on every screen transition. Don't understand why this behaviour is not considered as a bug. I have the latest version of all react-navigation packages. 

```
 "@react-navigation/drawer": "6.1.8",
    "@react-navigation/native": "6.0.6",
    "@react-navigation/stack": "6.0.11",
```

**Comment by shelzTimesignature** [comment]
> I had the same issue a while ago and solved the problem by setting the backgroundColor of react navigations theme to my custom backgroundColor

You didn't solve the problem man, you just don't see it . This problem is associated with a spring animation wch occurs when you move from one screen to the next.....

**Comment by shelzTimesignature** [comment]
> @cristianoccazinsp See my comment above. Adding a `theme` to the `NavigationContainer` fixes the issue.

Your solution only solved your problem man, try to change the header color to a different color wch is not white or any light colors close to white you will see that the problem persist

**Comment by shelzTimesignature** [comment]
headerMode="float",
animation:"fade"

Seem to solve the issue for me but I know this is a temporary fix

**Comment by dngconsulting** [comment]
For me, these options fixed definitively the issue, it's still a hack but that works 
React Navigation 6 : 
```
    screenOptions={{
        cardOverlayEnabled: false,
        animationEnabled: Platform.OS === 'ios',
        gestureEnabled: Platform.OS === 'ios',
        headerMode: 'float',
        presentation: 'transparentModal',
      }}
```

**Comment by seanhuggins1** [comment]
@aforty could you re-open this issue? It's a nasty bug that's lots of folks are having and it's easily reproducible. I created an issue on it awhile back as well https://github.com/react-navigation/react-navigation/issues/9891

`detachPreviousScreen: false` is the only good workaround as of yet but I am not sure of the performance implications of using that flag.

**Comment by JurieSpies** [comment]
`animationEnabled:false` is what did the trick for me on react-navigation 6

`<NavigationContainer>
      <Stack.Navigator
        screenOptions={{animationEnabled: false}}>`

**Comment by moherik** [comment]
In my case, i have GestureHandlerRootView wrap a NavigationContainer

`<GestureHandlerRootView style={{ flex: 1 }}><NavigationContainer theme={combinedTheme}>...</NavigationContainer></GestureHandlerRootView>`

Fixed it by, set backgroundColor of GestureHandlerRootView same as NavigationContainer color shceme 
` backgroundColor: theme.colors[colorScheme].background`

**Comment by mMarcos208** [comment]
for me `GestureHandlerRootView` was with white background color
